chainstate/src/detail/tests/mod.rs

```diff
@@ -47,6 +49,19 @@ const ERR_STORAGE_FAIL: &str = "Storage failure";
 const ERR_CREATE_BLOCK_FAIL: &str = "Creating block caused fail";
 const ERR_CREATE_TX_FAIL: &str = "Creating tx caused fail";
 
+#[must_use]
+fn make_seedable_rng(seed_opt: Option<&str>) -> impl Rng {
+    let seed = match seed_opt {
+        Some(s) => FromHex::from_hex(s).expect("seed is invalid"),
+        None => crypto::random::make_true_rng().gen::<[u8; 32]>(),
+    };
+    log::warn!(
```

Comment:
It is possible to specify separate log levels for specific modules. Something like that: `RUST_LOG="warn,test::foo=info,test::foo::bar=debug".

Reply:
Printing many things is OK. The right way to get this done is to be able to detect what test is being run, and then print that test name. Something like: `"Using seed with value 0xdeadbeef for test chainstate::abc::xyz::something_to_test()"`. Print that as info for every test. Then it's easy to find it if something goes wrong.

Reply:
I was thinking about println vs log approach. And I see one major benefit of printing - usability. 
If a test works we don't want to see any additional info and if it fails we want to see a seed to be able to hardcode it in the test to reproduce and fix it. This is exactly what println in tests does out of the box.
In the case of logging, the info level is disabled by default (otherwise every run it would flood the output with useless info). If a test fails - we see nothing, we rerun a particular test with proper log level, and boom ... new seed is generated.

Reply:
Honestly, I don't see benefits of "proper" logging in tests either. A typical log library gives many advantages such as automatically prepending current time to the output, flexibility (we can change the log level for every crate separately), a possibility to redirect output in a file, etc. None of this is needed for tests. Using `println!` is simple, doesn't require an initialization and as Gosha just mentioned Cargo already knows how to deal with it.

I think it resembles the situation with unwraps: we want to reduce their amount as much as possible and every usage is paired with a comment (usually in the form of expect). But it is totally OK to unwrap everything in the tests.

Reply:
I don't like using `println`, it feels too adhoc to me. `warn` is fine for me. But this isn't really the point of contention for me. I'm more concerned being able to see the test name printed.

Reply:
Now I print file and line using macro.
`warn` is not enough, the default level for out env_logger is Error, and logging seed like an error looks weird.  Printing seed with the rest of test output to std is much better imo.

Reply:
I feel like I'm missing some context here, but the output is already structured by tests. I have added an explicit panic in two chainstate tests, here is the output:
```
---- detail::tests::double_spend_tests::spend_output_in_the_same_block stdout ----
chainstate/src/detail/tests/double_spend_tests.rs:45 Using seed '3391698699346327995' for the PRNG
thread 'detail::tests::double_spend_tests::spend_output_in_the_same_block' panicked at 'not yet implemented', chainstate/src/detail/tests/double_spend_tests.rs:46:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order stdout ----
chainstate/src/detail/tests/double_spend_tests.rs:89 Using seed '6744204709038113292' for the PRNG
thread 'detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order' panicked at 'not yet implemented', chainstate/src/detail/tests/double_spend_tests.rs:94:9


failures:
    detail::tests::double_spend_tests::spend_output_in_the_same_block
    detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order
```
Therefore I see no reason in printing the file and the line in the `make_seedable_rng` macro. We will still be able to see the test name.

Reply:
This depends on the runner. Remember that the tests are allowed to run in parallel, so in general assuming order is wrong. Try `cargo nextest` for example. I'm sure this is gonna be a mess.

Reply:
Yes, but Cargo is already running tests in parallel and it works fine with it. I assume that output is captured independently, so there is no mess, unless the `--nocapture` flag is specified. I agree that different runners can potentially behave differently, but I think it would be strange to provide a worse user experience. As for `cargo-nextest` it produces the following output:

<details>
<summary>cargo-nextest output</summary>

```
--- STDOUT:                          chainstate detail::tests::double_spend_tests::double_spend_tx_in_another_block ---

running 1 test
chainstate/src/detail/tests/double_spend_tests.rs:191 Using seed '17993454279253376825' for the PRNG
test detail::tests::double_spend_tests::double_spend_tx_in_another_block ... FAILED

failures:

failures:
    detail::tests::double_spend_tests::double_spend_tx_in_another_block

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 52 filtered out; finished in 0.00s


--- STDERR:                          chainstate detail::tests::double_spend_tests::double_spend_tx_in_another_block ---
thread 'detail::tests::double_spend_tests::double_spend_tx_in_another_block' panicked at 'double_spend_tx_in_another_block', chainstate/src/detail/tests/double_spend_tests.rs:192:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure: 7 tests still running
        PASS [   0.009s]             chainstate detail::pow::work::tests::proof_of_work_ok_test
        FAIL [   0.006s]             chainstate detail::tests::double_spend_tests::double_spend_tx_in_the_same_block

--- STDOUT:                          chainstate detail::tests::double_spend_tests::double_spend_tx_in_the_same_block ---

running 1 test
chainstate/src/detail/tests/double_spend_tests.rs:137 Using seed '8590187921271618302' for the PRNG
test detail::tests::double_spend_tests::double_spend_tx_in_the_same_block ... FAILED

failures:

failures:
    detail::tests::double_spend_tests::double_spend_tx_in_the_same_block

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 52 filtered out; finished in 0.00s


--- STDERR:                          chainstate detail::tests::double_spend_tests::double_spend_tx_in_the_same_block ---
thread 'detail::tests::double_spend_tests::double_spend_tx_in_the_same_block' panicked at 'double_spend_tx_in_the_same_block', chainstate/src/detail/tests/double_spend_tests.rs:138:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

        PASS [   0.009s]             chainstate detail::tests::double_spend_tests::spend_output_in_the_same_block
        FAIL [   0.008s]             chainstate detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order

--- STDOUT:                          chainstate detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order ---

running 1 test
chainstate/src/detail/tests/double_spend_tests.rs:88 Using seed '17998526364629660100' for the PRNG
test detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order ... FAILED

failures:

failures:
    detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 52 filtered out; finished in 0.00s


--- STDERR:                          chainstate detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order ---
thread 'detail::tests::double_spend_tests::spend_output_in_the_same_block_invalid_order' panicked at 'spend_output_in_the_same_block_invalid_order', chainstate/src/detail/tests/double_spend_tests.rs:89:9
``` 
</details>


I would say the output is less clear than with the standard test runner, but it is still separated.